mempool/src/core_mempool/index.rs

```diff
@@ -255,6 +256,100 @@ impl TimelineIndex {
     }
 }
 
+pub struct MultiBucketTimelineIndex {
+    timelines: Vec<TimelineIndex>,
+    bucket_mins: Vec<u64>,
+}
+
+impl MultiBucketTimelineIndex {
+    pub(crate) fn new(bucket_mins: Vec<u64>) -> anyhow::Result<Self> {
+        anyhow::ensure!(!bucket_mins.is_empty(), "Must not be empty");
+        anyhow::ensure!(bucket_mins[0] == 0, "First bucket must start at 0");
+
+        let mut prev = None;
+        let mut timelines = vec![];
+        for entry in bucket_mins.clone() {
+            if let Some(prev) = prev {
+                anyhow::ensure!(prev < entry, "Values must be sorted and not repeat");
+            }
+            prev = Some(entry);
+            timelines.push(TimelineIndex::new());
+        }
+
+        Ok(Self {
+            timelines,
+            bucket_mins,
+        })
+    }
+
+    /// Read all transactions from the timeline since <timeline_id>.
+    /// At most `count` transactions will be returned.
+    pub(crate) fn read_timeline(
+        &self,
+        timeline_id: &TimelineId,
+        count: usize,
+    ) -> Vec<Vec<(AccountAddress, u64)>> {
+        assert!(timeline_id.0.len() == self.bucket_mins.len());
+
+        let mut added = 0;
+        let mut returned = vec![];
```

Comment:
nit: we can do `vec![vec![]; self.timelines.len()]` and use enumerate instead of zip

Reply:
We still need the zip for the timeline ids, so didn't change.

Reply:
we can use timeline_id[index] and avoid the while loop below?